specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,449 @@
+the `msg.sender` and `_newDelegations` parameters.
+
+At the end, the `delegate` function MUST emit a `DelegationCreated` event with the given function parameters.
+
+#### `delegateFromToken`
+
+Delegates 100% of the token voting power of an address (delegator) to another address (delegatee), mimicking the behavior
+of the `ERC20Votes`'s `delegate` function for backwards compatibility. This function MUST be callable only by the
+`GovernanceToken` contract as part of its `delegate` and `delegateBySig` functions.
+
+```solidity
+function delegateFromToken(address _delegator, address _delegatee) external
+```
+
+This function MUST enforce the migration logic, as specified in the [Migration](#migration) section, for the `_delegator`
+and `_delegatee` addresses. Afterwards, the function MUST call the [`_delegate`](#_delegate) with the `_delegator` and
+basic delegation parameters.
+
+At the end, the `delegateFromToken` function MUST emit a `DelegationCreated` event with the given function parameters.
+
+#### `delegateBatched`
+
+Delegates voting power to multiple addresses (delegatees) using the delegation array. This function is intended to be
+called by users.
+
+```solidity
+function delegateBatched(Delegation[] calldata _delegations) external
+```
+
+This function MUST enforce the migration logic, as specified in the [Migration](#migration) section, for `msg.sender`
+and every delegatee address in the `_delegations` array. Afterwards, the function MUST call the [`_delegate`](#_delegate)
+with the `msg.sender` and `_delegations` parameters.
+
+At the end, the `delegateBatched` function MUST emit a `DelegationsCreated` event with the given function parameters.
+
+#### `afterTokenTransfer`
+
+Updates the voting power of two addresses (`_from` and `_to`) after a token transfer. This function MUST only be callable
+by the `GovernanceToken` contract.
+
+```solidity
+function afterTokenTransfer(address _from, address _to, uint256 _amount) external
+```
+
+This function MUST enforce the migration logic, as specified in the [Migration](#migration) section, for the `_from`
+and `_to` addresses.
+
+If the token transfer is due to a mint or burn operation, the function MUST update the total supply checkpoints with
+the new total supply (subtracting `_amount` for a burn operation and adding `_amount` for a mint operation). Afterwards,
+the function MUST calculate and update the voting power of the delegations of `_from` and `_to`. This step MUST apply
+the voting power adjustments to the checkpoints of the `_from`, `_to`, and delegatee addresses.
+
+#### `migrateAccounts`
+
+Migrates the delegation state of the given accounts from the `GovernanceToken` to the `GovernanceDelegation` contract. This
+function MUST iterate over the list of `_accounts` addresses and apply the logic specified in the
+[Migration](#migration) section.
+
+```solidity
+function migrateAccounts(address[] calldata _accounts) external
+```
+
+#### `_delegate`
```

Comment:
Yeah basic feedback on this for future ref would just be that the spec should really cover external interface and expected function behavior. In some sense what this section is really doing is providing common function behavior for a number of external functions. Readers of the spec shouldn't really care that those functions specifically call an internal function called `_delegate` but they *do* care that those external functions implement the function behavior being described here. I think this is perfectly fine for now but in a future spec cleanup we can probably streamline this a bit.

Reply:
Agreed with this, defining an internal function like this is in the weeds with implementation details